beacon-chain/core/altair/transition.go

```diff
@@ -103,3 +115,103 @@ func ProcessEpoch(ctx context.Context, state iface.BeaconStateAltair) (iface.Bea
 
 	return state, nil
 }
+
+// ProcessSlots process through skip slots and apply epoch transition when it's needed.
+//
+// Spec pseudocode definition:
+//  def process_slots(state: BeaconState, slot: Slot) -> None:
+//    assert state.slot < slot
+//    while state.slot < slot:
+//        process_slot(state)
+//        # Process epoch on the start slot of the next epoch
+//        if (state.slot + 1) % SLOTS_PER_EPOCH == 0:
+//            process_epoch(state)
+//        state.slot = Slot(state.slot + 1)
+func ProcessSlots(ctx context.Context, state iface.BeaconState, slot types.Slot) (iface.BeaconState, error) {
+	ctx, span := trace.StartSpan(ctx, "altair.ProcessSlots")
+	defer span.End()
+	if state == nil {
+		return nil, errors.New("nil state")
+	}
+	span.AddAttributes(trace.Int64Attribute("slots", int64(slot)-int64(state.Slot())))
+
+	// The block must have a higher slot than parent state.
+	if state.Slot() >= slot {
+		err := fmt.Errorf("expected state.slot %d < slot %d", state.Slot(), slot)
+		traceutil.AnnotateError(span, err)
+		return nil, err
+	}
+
+	highestSlot := state.Slot()
+	key, err := coreState.CacheKey(ctx, state)
+	if err != nil {
+		return nil, err
+	}
+
+	// Restart from cached value, if one exists.
+	cachedState, err := SkipSlotCache.Get(ctx, key)
+	if err != nil {
+		return nil, err
+	}
+
+	if cachedState != nil && cachedState.Slot() < slot {
+		highestSlot = cachedState.Slot()
+		state = cachedState
+	}
+	if err := SkipSlotCache.MarkInProgress(key); errors.Is(err, cache.ErrAlreadyInProgress) {
+		cachedState, err = SkipSlotCache.Get(ctx, key)
+		if err != nil {
+			return nil, err
+		}
+		if cachedState != nil && cachedState.Slot() < slot {
+			highestSlot = cachedState.Slot()
+			state = cachedState
+		}
```

Comment:
Don't we already have the very same values for `highestSlot` and `state`? We obtained values from cache, then tried to mark key as in progress, and if it is already in progress, we run the very same code to populate `highestSlot` and `state`. That's kind of confusing.

Reply:
A little. We can generally revisit this outside of this PR. It's been a while and last time I checked this cache did have good hit rate during long finality